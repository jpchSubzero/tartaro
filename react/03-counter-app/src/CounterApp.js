//Importar el hook useState para manejar el estado
import React, { useState } from 'react';
import PropTypes from 'prop-types';

//Snippet -> rafce función con exportación default
//Snippet -> rafcp función con exportación default y propTypes
//Snippet -> rafc función con exportación definida
//Eventos sintéticos (SyntheticEvent) https://es.reactjs.org/docs/events.html
// const CounterApp = ({value}) => {

//Poner un valor por defecto para que funcione siempre
const CounterApp = ({value=0}) => {

    //Desestructurar la respuesta de useState que es una variable y una función, por convención se utiliza [<Variable>, set<Variable>]. <Variable> es inmutable aunque usemos let en lugar de const
    const [valor, setValor] = useState(value);

    //Por convención se utiliza handle como prefijo en una función que maneja un evento
    const handleAdd = () => {
        //No se puede utilizar ++ porque los valores de estado son inmutables y no pueden modificarse directamente
        // valor++;
        // setValor(valor++);

        //Para modificar se utilizar la función definida en el hook (setValor) y un nuevo valor
        setValor(valor + 1);

        //Si no se tuviera acceso a <Variable> se puede modificar a través de una función anónima porque set<Variable> sabe que esta ligada a <Variable> y se puede recibir como parámetro y luego incrementarla
        // setValor((c) => c + 1);

        // console.log(valor);
    }

    // //Ejemplo recibiendo el evento como parámetro
    // const handleAdd = (e) => {
    //     value++;
    //     console.log(value);
    // }

    //Ejemplo de función que devuelve una función
    // const handleAdd1 = () => {
    //     return () => console.log('Respuesta como función');
    // }

    //#region Tarea 2
    //Crear método para reducir contador en 1, otro para resetear al value que llegó y definirle un valor por defecto en caso de que no llegue un value por parámetro

    //Función resumida porque tiene una sola instrucción
    const handleSubstract = () => setValor(valor - 1);
    
    //Función resumida porque tiene una sola instrucción
    const handleReset = () => setValor(value);

    //#endregion

    return (
        <>
            <h1>CounterApp</h1>
            
            {/* Mostrar el valor que llega como parámetro */}
            {/* <h2> { value } </h2>   
             */}

             {/* Mostrar el valor como estado del componente que se puede modificar con setValor creado con el hook useState */}
            <h2> { valor } </h2>   
            
            {/* Mostrar el detalle del evento */}
            {/* <button onClick={ (e) => { console.log(e) } }>+1</button> */}
            
            {/* Invocar una función a través de un evento, forma larga */}
            {/* <button onClick={ (e) => handleAdd(e) }>+1</button> */}
            
            {/* Invocar una función a través de un evento, forma corta basada en que JS asume que la función que recibe un parámetro (e) lo pasa como parámetro a la función que contiene (handleAdd)*/}
            <button onClick={ handleAdd }>+1</button>
            
            {/* No debe invocarse con (), handleAdd(), porque react la ejecuta sin necesidad del botón, como si fuera una instrucción. Se necesita cuando la función no devuelve un valor sino otra función*/}
            {/* <button onClick={ handleAdd1() }>+1</button> */}

             {/* Tarea 2 */}
            <button onClick={ handleSubstract }>-1</button>
            <button onClick={ handleReset }>Reset</button>

        </>
    );
}

CounterApp.propTypes = {
    value:PropTypes.number
}

export default CounterApp;