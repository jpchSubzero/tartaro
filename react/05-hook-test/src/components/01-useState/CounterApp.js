import React, { useState } from 'react';
// Agregar css
import './counter.css';

export const CounterApp = () => {

    // Crear el useState
    // const [counter, setCounter] = useState(10);

    // Cambiar el entero por un objeto
    // const [counter, setCounter] = useState({
    //     counter1: 10,
    //     counter2: 20
    // });

    // Desestructurar el state para no utilizar . en la referencia
    // const [{ counter1, counter2 }, setCounter] = useState({

    // Para no tener que hacer referencias a cada propiedad es mejor no destructurar en la definición del useState
    const [state, setState] = useState({
        counter1: 10,
        counter2: 20,

        // Aumentar para revisar el problema de tener muchas propiedades y la ineficiencia de código de repetirlas
        counter3: 30,
        counter4: 40,
        counter5: 50,
        counter6: 60
    });

    // Destructurar el state con las propiedades que necesitemos utilizar
    const {counter1, counter2} = state;

    return (
        <>
            {/* <h1>Counter { 0 }</h1> */}

            {/* Cambiar el dato quemado por el state */}
            {/* <h1>Counter { counter }</h1> */}

            {/* Hacer referencia a los objetos en el state */}
            {/* <h1>Counter1 { counter.counter1 }</h1>
            <h1>Counter2 { counter.counter2 }</h1> */}

            {/* Obtener desde el state destructurado */}
            <h1>Counter1 { counter1 }</h1>
            <h1>Counter2 { counter2 }</h1>

            <hr />

            {/* <button className="btn btn-primary">+1</button> */}

            {/* Implementar el evento click */}
            <button className="btn btn-primary" onClick={ () => {
                // Comentar porque ya no existe
                // setCounter(counter + 1);
                
                // Esto incrementa el counter1 pero como es una definición de estado también elimina las propiedades que no consten, es decir, counter2
                // setCounter({
                //     counter1: counter1 + 1
                // });

                // Para solucionar se puede asignar el resto de propiedades. El problema vine cuando tenemos muchas propiedades y no es óptimo el código
                // setCounter({
                //     counter1: counter1 + 1,
                //     counter2: counter2
                // });

                // Destructurar el state y solo asignar valores a las propiedades que necesitemos, de esta manera no se pierden las propiedades anteriores
                setState({
                    ...state,
                    counter1: counter1 + 1
                });                
            }}>+1</button>
        </>
    )
}

// Tarea 1
// Incrementar solo el counter 1 al hacer click en +1