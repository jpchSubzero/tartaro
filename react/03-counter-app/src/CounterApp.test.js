//Configurar el index para que renderice CounterApp y verificar

import React from 'react';
import CounterApp from "./CounterApp";
import { shallow } from 'enzyme';
import '@testing-library/jest-dom';

//Definir para no repetir en cada test
// const wrapper = shallow(<CounterApp />);

//Cambiar a let para poder reinicializar
let wrapper = shallow(<CounterApp />);

describe('Pruebas del archivo CounterApp', () => {
    //Reinicializar variables antes de cada test
    beforeEach(() => {
        wrapper = shallow(<CounterApp />);
    });

    //Validar el componente
    test('Debe presentar correctamente CounterApp', () => {
        // const wrapper = shallow(<CounterApp />);

        expect(wrapper).toMatchSnapshot();
    });

    //Validar información en compoonentes html
    test('Debe mostrar el value enviado en props', () => {
        const value = 100;

        const wrapper = shallow(<CounterApp value = { value }/>);

        const textovalue = parseInt(wrapper.find('h2').text().trim());
        console.log(textovalue);

        expect(textovalue).toBe(value);
    });

    //Eventos click
    test('Debe incrementar el contador con el botón +1', () => {
        // const wrapper = shallow(<CounterApp />);

        //Buscar botones y de ellos el primero que es el +1
        // const botonAdd = wrapper.find('button').at(0);
        // console.log(botonAdd.html());

        wrapper.find('button').at(0).simulate('click');
        const counterTexto = wrapper.find('h2').text().trim();
        //Como no se pasa parámetro el value por defecto es 0 y al hacer click se hace 1
        expect(counterTexto).toBe('1');
    })

    //Tarea, hacer con botón -1
    test('Debe decrementar el contador con el botón -1', () => {
        wrapper.find('button').at(1).simulate('click');
        const counterTexto = wrapper.find('h2').text().trim();

        //Solución 1
        //Como no se pasa parámetro el value por defecto es 0 y al hacer click se hace -1 pero como se ejecuta el click en la prueba anterior se inicia en 1
        // expect(counterTexto).toBe('0');

        //Solución 2
        //Utilizar beforeEach() que ejecuta algo antes de cada test
        expect(counterTexto).toBe('-1');

    })
    
    //Tarea, hacer con botón reset
    test('Debe resetear el contador al value', () => {
        //Agregar acciones previas para saber si vuelve después de cambios, sino no se puede saber y funciona o no
        wrapper.find('button').at(0).simulate('click');
        wrapper.find('button').at(0).simulate('click');
        wrapper.find('button').at(0).simulate('click');
        wrapper.find('button').at(0).simulate('click');
        wrapper.find('button').at(0).simulate('click');
        wrapper.find('button').at(0).simulate('click');
        wrapper.find('button').at(0).simulate('click');
        wrapper.find('button').at(1).simulate('click');
        wrapper.find('button').at(1).simulate('click');

        console.log(wrapper.find('h2').text().trim());

        wrapper.find('button').at(2).simulate('click');
        const counterTexto = wrapper.find('h2').text().trim();

        expect(counterTexto).toBe('0');

    })    
})
