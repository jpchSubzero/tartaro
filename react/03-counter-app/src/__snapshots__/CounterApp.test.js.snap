// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del archivo CounterApp Debe presentar correctamente CounterApp 1`] = `
<Fragment>
  <h1>
    CounterApp
  </h1>
  <h2>
     
    0
     
  </h2>
  <button
    onClick={[Function]}
  >
    +1
  </button>
  <button
    onClick={[Function]}
  >
    -1
  </button>
  <button
    onClick={[Function]}
  >
    Reset
  </button>
</Fragment>
`;
